File: python/cleaning_setting_sys.py
import os
from enum import Enum, auto

FILE_PATH = os.path.join(os.getcwd(),"values.env")

def decorator_for_show_func (func) :
  def wrapper (*args, **kwargs) :
    print()
    func(*args, **kwargs)
  
  return wrapper

def str_form (str_list) :
  length_of_str_list = len(str_list)
  padding_size = [10 for _ in range(length_of_str_list - 1)]
  padding_size.insert(0, 20)
  str_format = ""
  for idx in range(length_of_str_list) :
    each = str_list[idx]
    str_format += each + ' ' * (padding_size[idx] - calculate_length_of_str(each))
  str_format += '\n'

  return str_format

def calculate_length_of_str (str_input) :
  length_of_str_input = 0
  for char in str_input : 
    if '\uAC00' <= char <= '\uD7AF' :
      length_of_str_input += 2
    else : 
      length_of_str_input += 1
  
  return length_of_str_input


class Status (Enum) :
  normal = auto()

  cleaning_room = "생활관 청소"
  seperating_trash_each_room = "생활관 분리수거"
  seperating_trash = "분리수거"

  vacation = "휴가"
  vacation_but_coming_this_week = "휴가복귀"
  dispatch = "파견"
  discharge = "전역"
  sleepover = "외박"


class RoomMembers :
    
  def __init__ (self, file_path) :
    self.file_contents = self.read_file_contents(file_path)
    self.members_status = {}
    self.members_map = self.get_members_map()

    self.excluded_members = self.members_map["excluded_members"]
    self.excluded_but_lottery_members = self.members_map["excluded_but_lottery_members"]
    self.cleaning_members_before_week = self.members_map["cleaning_members_before_week"]

    self.cleaning_members = {}
    self.cleaning_members_list = []

  def read_file_contents(self, file_path) :
    with open(file_path, 'r') as file :
      contents = file.readlines()
    return contents
  
  def get_members_map (self) :
    members_map = {}

    for idx in range(4) : 
      room_number = "room" + str(idx + 1)
      members_map[room_number] = [i.strip() for i in self.file_contents[idx].split(':')[1].split(',')]
      self.set_status(members_map[room_number], Status.normal)

    raw_excluded_members = ""

    special_reasoned_members = []
    for idx in range(4) : 
      raw_members_tmp = self.file_contents[idx + 11].split(':')[1]
      special_reasoned_members.insert(idx, raw_members_tmp)
      if raw_members_tmp.strip() != '' : 
        raw_excluded_members += raw_members_tmp + ','
        
    vacation_members = special_reasoned_members[0]
    dispatch_members = special_reasoned_members[1]
    discharge_members = special_reasoned_members[2]
    sleepover_members = special_reasoned_members[3]

    self.set_status(vacation_members, Status.vacation)
    self.set_status(dispatch_members, Status.dispatch)
    self.set_status(discharge_members, Status.discharge)
    self.set_status(sleepover_members, Status.sleepover)

    
    members_map["seperating_trash"] = {}
    for idx in range(4) : 
      each_excluded_members = self.file_contents[idx + 16].split(':')[1].strip()
      members_map["seperating_trash"]["room" + str(idx + 1)] = each_excluded_members
      raw_excluded_members += each_excluded_members + ',' 

      self.members_status[each_excluded_members] = Status.seperating_trash
    
    members_map["excluded_members"] = [i.strip() for i in raw_excluded_members[:-1].split(',')]
    members_map["excluded_but_lottery_members"] = [i.strip() for i in self.file_contents[21].split(':')[1].split(',')]
    for each in members_map["excluded_but_lottery_members"] :
      self.set_status(each, Status.vacation_but_coming_this_week)

    members_map["cleaning_members_before_week"] = {}
    for idx in range(4) :
      members_map["cleaning_members_before_week"]["room" + str(idx + 1)] = { "cleaning_room" : self.file_contents[idx + 6].split(':')[1].split(',')[0].strip(), "seperating_trash_each_room" : self.file_contents[idx + 6].split(':')[1].split(',')[1].strip() }

    return members_map

  def set_status (self, member, status) :
    if type(member) is list : 
      for each in member :
        if each.strip() != '' :
          self.members_status[each] = status
    else :
      if type(member) is str :
        member = member.split(',')
        if type(member) is list :
          for each in member :
            if each.strip() != '' :
              self.members_status[each.strip()] = status
        else :
          member = member.strip()
          if member != '' :
            self.members_status[member] = status

  def get_next_cleaning_members (self) :
    cleaning_members = {}
    for idx in range(4) :
      room_number = "room" + str(idx + 1)
      cleaning_members[room_number] = {}
      cleaning_members[room_number] = self.get_next_members(room_number)

      cleaning_members[room_number]["seperating_trash"] = self.members_map["seperating_trash"][room_number]

    self.cleaning_members = cleaning_members
    self.cleaning_members_list = [val for each in cleaning_members.values() for val in each.values()]
    return cleaning_members

  def get_next_members (self, room_number) : 
    room_members = self.members_map[room_number]
    room_members_length = len(room_members)

    before_work_member = self.members_map["cleaning_members_before_week"][room_number]
    before_cleaning_room_member = before_work_member["cleaning_room"]
    before_seperating_trash_each_room_member = before_work_member["seperating_trash_each_room"]

    excluded_members = self.excluded_members
    before_cleaning_room_member_idx = room_members.index(before_cleaning_room_member)
    before_seperating_trash_each_room_member_idx = room_members.index(before_seperating_trash_each_room_member)

    next_cleaning_room_member = None
    count = 0
    idx = before_cleaning_room_member_idx + 1
    if idx == room_members_length : 
      idx = 0

    while(not count == room_members_length) :
      this_member = room_members[idx]
      if this_member not in excluded_members :
        if next_cleaning_room_member is None : 
          next_cleaning_room_member = this_member
        else :
          if this_member != before_seperating_trash_each_room_member :
            self.members_status[next_cleaning_room_member] = Status.cleaning_room
            self.members_status[this_member] = Status.seperating_trash_each_room
            return {"cleaning_room" : next_cleaning_room_member, "seperating_trash_each_room" : this_member}

      if (idx + 1) == room_members_length :
        idx = 0
      else :
        idx += 1
      count += 1

    return "error"

  def get_lottery_members (self) :
    lottery_members = {}
    for idx in range(4) :
      room_number = "room" + str(idx + 1)
      lottery_members[room_number] = self.get_lottery_members_each_room(room_number)

    return lottery_members

  def get_lottery_members_each_room (self, room_number) :
    room_members = self.members_map[room_number]
    excluded_members = self.members_map["excluded_members"]
    excluded_but_lottery_members = self.members_map["excluded_but_lottery_members"]
    cleaning_members_list = self.cleaning_members_list

    return list(filter(lambda each : each not in cleaning_members_list and ((each in excluded_members) is (each in excluded_but_lottery_members)) , room_members))

  @decorator_for_show_func
  def show_status (self) :
    print("전체인원 현황")
    for idx in range(4) :
      room_number = "room" + str(idx + 1)
      before_cleaning_members = self.cleaning_members_before_week[room_number]
      members = self.members_map[room_number]
      print(str(idx + 1) + "생활관 : ", end = '')
      for each in members :
        status = self.members_status[each]
        print(each, end = '')

        if each == before_cleaning_members["cleaning_room"] : 
          print("(이전 생활관 청소)", end = '')
        if each == before_cleaning_members["seperating_trash_each_room"] :
          print("(이전 생활관 분리수거)", end = '')

        if status == Status.normal :
          print(', ', end = '')
        else :
          print('(' + self.members_status[each].value + "), ", end = '')
      print(end = "\n\n")
    
  def get_before_week_cleaning_members_by_area (self) :
    cleaning_members_before_week = self.cleaning_members_before_week
    before_week_cleaning_members_by_area = {}
    for idx in range(4) :
      room_number = "room" + str(idx + 1)
      before_week_cleaning_members_by_area["cleaning_room"] = cleaning_members_before_week[room_number]["cleaning_room"]
      before_week_cleaning_members_by_area["seperating_trash_each_room"] = cleaning_members_before_week[room_number]["seperating_trash_each_room"]

    return before_week_cleaning_members_by_area

  @decorator_for_show_func
  def show_summary (self) :
    room_numbers = [str(idx + 1) + "생활관" for idx in range(4)]
    room_numbers.insert(0, '')

    cleaning_room = ["생활관 청소"]
    seperating_trash_each_room = ["생활관 분리수거"]
    seperating_trash = ["분리수거"]
    cleaning_members = self.get_next_cleaning_members()
    for idx in range(4) :
      room_members = cleaning_members["room" + str(idx + 1)]
      cleaning_room.insert(idx + 1, room_members["cleaning_room"])
      seperating_trash_each_room.insert(idx + 1, room_members["seperating_trash_each_room"])
      seperating_trash.insert(idx + 1, room_members["seperating_trash"])

    str_output = ""
    str_output += str_form(room_numbers)
    str_output += str_form(cleaning_room)
    str_output += str_form(seperating_trash_each_room)
    str_output += str_form(seperating_trash)

    print(str_output)

  @decorator_for_show_func
  def show_lottery_members (self) :
    lottery_members = self.get_lottery_members()

    print("제비뽑기 인원")
    for idx in range(4) :
      room_number = "room" + str(idx + 1)
      print(str(idx + 1) + "생활관 : " +  str(lottery_members[room_number]))


if __name__ == "__main__" :
  room_members = RoomMembers(FILE_PATH)

  room_members.show_summary()
  room_members.show_lottery_members()
  room_members.show_status()

File: dart/cleaning_setting_sys.dart
import 'dart:io';

const String INPUT_FILE_PATH = 'values.env';
const String OUTPUT_FILE_PATH = 'output.env';

typedef void ShowFunc();

void decoratorForShowFunc(ShowFunc func) {
  print('');
  func();
}

String strForm(List<String> strList) {
  int lengthOfStrList = strList.length;
  List<int> paddingSize = List<int>.filled(lengthOfStrList - 1, 10);
  paddingSize.insert(0, 20);
  String strFormat = '';
  for (int idx = 0; idx < lengthOfStrList; idx++) {
    String each = strList[idx];
    strFormat += each + ' ' * (paddingSize[idx] - calculateLengthOfStr(each));
  }
  strFormat += '\n';

  return strFormat;
}

int calculateLengthOfStr(String strInput) {
  int lengthOfStrInput = 0;
  for (int i = 0; i < strInput.length; i++) {
    String char = strInput[i];
    if ('\uAC00' <= char && char <= '\uD7AF') {
      lengthOfStrInput += 2;
    } else {
      lengthOfStrInput += 1;
    }
  }

  return lengthOfStrInput;
}

enum Status {
  normal,
  cleaningRoom,
  seperatingTrashEachRoom,
  seperatingTrash,
  vacation,
  vacationButComingThisWeek,
  dispatch,
  discharge,
  sleepover,
}

class RoomMembers {
  late List<String> fileContents;
  Map<String, dynamic> membersStatus = {};
  late Map<String, dynamic> membersMap;
  late List<String> excludedMembers;
  late List<String> excludedButLotteryMembers;
  late Map<String, dynamic> cleaningMembersBeforeWeek;
  late Map<String, dynamic> cleaningMembers;
  late List<dynamic> cleaningMembersList;

  RoomMembers(String filePath) {
    fileContents = readFileContents(filePath);
    membersMap = getMembersMap();
    excludedMembers = membersMap['excluded_members'];
    excludedButLotteryMembers = membersMap['excluded_but_lottery_members'];
    cleaningMembersBeforeWeek = membersMap['cleaning_members_before_week'];
    cleaningMembers = {};
    cleaningMembersList = [];

    getMembersStatus(membersMap);
  }

  List<String> readFileContents(String filePath) {
    File file = File(filePath);
    return file.readAsLinesSync();
  }

  void writeFileContents(String contents, String filePath) {
    File file = File(filePath);
    file.writeAsStringSync(contents);
  }

  Map<String, dynamic> getMembersMap() {
    Map<String, dynamic> membersMap = {};

    for (int idx = 0; idx < 4; idx++) {
      String roomNumber = 'room${idx + 1}';
      membersMap[roomNumber] =
          fileContents[idx].split(':')[1].split(',').map((e) => e.trim()).toList();
      setMembersStatus(membersMap[roomNumber], Status.normal);
    }

    String rawExcludedMembers = '';

    List<String> specialReasonedMembers = [];
    for (int idx = 0; idx < 4; idx++) {
      String rawMembersTmp = fileContents[idx + 11].split(':')[1];
      specialReasonedMembers.insert(idx, rawMembersTmp);
      if (rawMembersTmp.trim() != '') {
        rawExcludedMembers += rawMembersTmp + ',';
      }
    }

    String vacationMembers = specialReasonedMembers[0];
    String dispatchMembers = specialReasonedMembers[1];
    String dischargeMembers = specialReasonedMembers[2];
    String sleepoverMembers = specialReasonedMembers[3];

    membersMap['vacation_members'] = vacationMembers.split(',').map((e) => e.trim()).toList();
    membersMap['dispatch_members'] = dispatchMembers.split(',').map((e) => e.trim()).toList();
    membersMap['discharge_members'] = dischargeMembers.split(',').map((e) => e.trim()).toList();
    membersMap['sleepover_members'] = sleepoverMembers.split(',').map((e) => e.trim()).toList();
    membersMap['excluded_members'] = rawExcludedMembers.split(',').map((e) => e.trim()).toList();
    membersMap['excluded_but_lottery_members'] = membersMap['excluded_members']
        .where((element) => !excludedButLotteryMembers.contains(element))
        .toList();
    return membersMap;
  }

  void setMembersStatus(List<String> members, Status status) {
    for (String member in members) {
      membersStatus[member] = status;
    }
  }

  void getMembersStatus(Map<String, dynamic> membersMap) {
    for (String room in membersMap.keys) {
      setMembersStatus(membersMap[room], Status.normal);
    }

    setMembersStatus(membersMap['vacation_members'], Status.vacation);
    setMembersStatus(membersMap['dispatch_members'], Status.dispatch);
    setMembersStatus(membersMap['discharge_members'], Status.discharge);
    setMembersStatus(membersMap['sleepover_members'], Status.sleepover);
  }

  void printMembersStatus() {
    List<String> membersStatusList = [];
    membersStatus.forEach((key, value) {
      membersStatusList.add('$key: $value');
    });

    print(membersStatusList.join('\n'));
  }

  void calculateCleaningMembers() {
    DateTime now = DateTime.now();
    DateTime oneWeekAgo = now.subtract(Duration(days: 7));
    cleaningMembersBeforeWeek.forEach((member, lastCleaningDate) {
      DateTime lastDate = DateTime.parse(lastCleaningDate);
      if (lastDate.isBefore(oneWeekAgo)) {
        cleaningMembers[member] = lastCleaningDate;
        cleaningMembersList.add(member);
      }
    });
  }

  String getNextCleaningMember() {
    if (cleaningMembersList.isEmpty) {
      calculateCleaningMembers();
    }

    if (cleaningMembersList.isEmpty) {
      return 'No member needs to clean this week.';
    } else {
      String nextMember = cleaningMembersList.first;
      cleaningMembersList.removeAt(0);
      return nextMember;
    }
  }

  void showRoomStatus() {
    print('Room Members:');
    membersMap.forEach((room, members) {
      print('Room $room: ${members.join(', ')}');
    });

    decoratorForShowFunc(() {
      print('Excluded Members: ${excludedMembers.join(', ')}');
      print('Excluded but Lottery Members: ${excludedButLotteryMembers.join(', ')}');
    });

    decoratorForShowFunc(() {
      print('Cleaning Members Before Week:');
      cleaningMembersBeforeWeek.forEach((member, lastCleaningDate) {
        print('$member: $lastCleaningDate');
      });
    });

    decoratorForShowFunc(() {
      print('Cleaning Members:');
      cleaningMembers.forEach((member, lastCleaningDate) {
        print('$member: $lastCleaningDate');
      });
    });

    decoratorForShowFunc(() {
      print('Members Status:');
      printMembersStatus();
    });
  }

  void saveMembersStatus() {
    List<String> outputLines = [];
    membersMap.forEach((room, members) {
      outputLines.add('Room $room: ${members.join(',')}');
    });

    decoratorForSaveFunc(() {
      outputLines.add('Excluded Members: ${excludedMembers.join(',')}');
      outputLines.add('Excluded but Lottery Members: ${excludedButLotteryMembers.join(',')}');
    });

    decoratorForSaveFunc(() {
      outputLines.add('Cleaning Members Before Week:');
      cleaningMembersBeforeWeek.forEach((member, lastCleaningDate) {
        outputLines.add('$member: $lastCleaningDate');
      });
    });

    decoratorForSaveFunc(() {
      outputLines.add('Cleaning Members:');
      cleaningMembers.forEach((member, lastCleaningDate) {
        outputLines.add('$member: $lastCleaningDate');
      });
    });

    decoratorForSaveFunc(() {
      outputLines.add('Members Status:');
      membersStatus.forEach((member, status) {
        outputLines.add('$member: $status');
      });
    });

    File('members_status.txt').writeAsStringSync(outputLines.join('\n'));
    print('Members status saved to members_status.txt.');
  }
}

void main() {
  print('===== Welcome to Room Management System! =====');
  RoomManagement roomManagement = RoomManagement();

  bool exitFlag = false;
  while (!exitFlag) {
    print('\nPlease select an action:');
    print('1. Show room status');
    print('2. Assign cleaning member');
    print('3. Save members status');
    print('4. Exit');

    String input = stdin.readLineSync();
    switch (input) {
      case '1':
        roomManagement.showRoomStatus();
        break;
      case '2':
        String nextCleaningMember = roomManagement.getNextCleaningMember();
        print('The next cleaning member is: $nextCleaningMember');
        break;
      case '3':
        roomManagement.saveMembersStatus();
        break;
      case '4':
        exitFlag = true;
        break;
      default:
        print('Invalid input. Please try again.');
    }
  }

  print('Thank you for using the Room Management System. Goodbye!');
}
